File: src/tensor.py
from __future__ import annotations
from enum import Enum
import numpy as np


class GradFn(Enum):
    ADD = "add"
    SUB = "sub"
    MUL = "mul"
    DIV = "div"
    POW = "pow"
    MEAN = "mean"
    EXP = "exp"
    LOG = "log"
    RELU = "relu"
    SIGMOID = "sigmoid"


class Tensor:
    def __init__(
        self,
        data: int | float | list | np.ndarray,
        requires_grad: bool = False,
        grad_fn: GradFn | None = None,
        parents: list[Tensor] | None = None,
    ):
        self.data = np.array(data, dtype=float)
        self.requires_grad = requires_grad
        self.grad_fn = grad_fn
        self.grad = np.zeros_like(self.data) if requires_grad else None
        self.parents = parents or []

    # ---------------- Utility ---------------- #
    def __match_shape(self, grad, shape):
        while len(grad.shape) > len(shape):
            grad = grad.sum(axis=0)
        for axis, (g, s) in enumerate(zip(grad.shape, shape)):
            if s == 1 and g > 1:
                grad = grad.sum(axis=axis, keepdims=True)
        return grad.reshape(shape)

    def __ensure_tensor(self, other: int | float | list | np.ndarray | Tensor) -> Tensor:
        if isinstance(other, Tensor):
            return other
        if type(other) in [int, float, list, np.ndarray]:
            return Tensor(other)
        raise TypeError(
            f"Expected Tensor, int, float, list or np.ndarray, got {type(other)}"
        )

    # ---------------- Operators ---------------- #
    def __add__(self, other: Tensor | int | float):
        other = self.__ensure_tensor(other)
        return Tensor(
            self.data + other.data, requires_grad=True,
            grad_fn=GradFn.ADD, parents=[self, other]
        )

    def __sub__(self, other: Tensor | int | float):
        other = self.__ensure_tensor(other)
        return Tensor(
            self.data - other.data, requires_grad=True,
            grad_fn=GradFn.SUB, parents=[self, other]
        )

    def __mul__(self, other: Tensor | int | float):
        other = self.__ensure_tensor(other)
        return Tensor(
            self.data * other.data, requires_grad=True,
            grad_fn=GradFn.MUL, parents=[self, other]
        )

    def __truediv__(self, other: Tensor | int | float):
        other = self.__ensure_tensor(other)
        return Tensor(
            self.data / other.data, requires_grad=True,
            grad_fn=GradFn.DIV, parents=[self, other]
        )

    def __pow__(self, other: Tensor | int | float):
        other = self.__ensure_tensor(other)
        return Tensor(
            self.data ** other.data, requires_grad=True,
            grad_fn=GradFn.POW, parents=[self, other]
        )

    # ---------------- Functions ---------------- #
    def mean(self):
        return Tensor(
            np.mean(self.data), requires_grad=True,
            grad_fn=GradFn.MEAN, parents=[self]
        )

    def exp(self):
        return Tensor(
            np.exp(self.data), requires_grad=True,
            grad_fn=GradFn.EXP, parents=[self]
        )

    def log(self):
        return Tensor(
            np.log(self.data), requires_grad=True,
            grad_fn=GradFn.LOG, parents=[self]
        )

    def relu(self):
        return Tensor(
            np.maximum(0, self.data), requires_grad=True,
            grad_fn=GradFn.RELU, parents=[self]
        )

    def sigmoid(self):
        sig = 1 / (1 + np.exp(-self.data))
        return Tensor(
            sig, requires_grad=True,
            grad_fn=GradFn.SIGMOID, parents=[self]
        )

    # ---------------- Autograd ---------------- #
    def zero_grad(self):
        if self.grad is not None:
            self.grad = np.zeros_like(self.data)

    def backward(self, grad: np.ndarray | float = 1.0):
        grad = np.array(grad, dtype=float)

        if self.requires_grad:
            grad = self.__match_shape(grad, self.data.shape)
            if self.grad is None:
                self.grad = grad
            else:
                self.grad += grad

        if self.grad_fn == GradFn.ADD:
            self.parents[0].backward(grad)
            self.parents[1].backward(grad)

        elif self.grad_fn == GradFn.SUB:
            self.parents[0].backward(grad)
            self.parents[1].backward(-grad)

        elif self.grad_fn == GradFn.MUL:
            grad_x = grad * self.parents[1].data
            grad_y = grad * self.parents[0].data
            self.parents[0].backward(self.__match_shape(
                grad_x, self.parents[0].data.shape))
            self.parents[1].backward(self.__match_shape(
                grad_y, self.parents[1].data.shape))

        elif self.grad_fn == GradFn.DIV:
            grad_x = grad / self.parents[1].data
            grad_y = -grad * self.parents[0].data / (self.parents[1].data ** 2)
            self.parents[0].backward(self.__match_shape(
                grad_x, self.parents[0].data.shape))
            self.parents[1].backward(self.__match_shape(
                grad_y, self.parents[1].data.shape))

        elif self.grad_fn == GradFn.POW:
            base, exp = self.parents
            safe_log = np.where(base.data > 0, np.log(base.data), 0.0)
            grad_base = grad * exp.data * (base.data ** (exp.data - 1))
            grad_exp = grad * (base.data ** exp.data) * safe_log
            self.parents[0].backward(
                self.__match_shape(grad_base, base.data.shape))
            self.parents[1].backward(
                self.__match_shape(grad_exp, exp.data.shape))

        elif self.grad_fn == GradFn.MEAN:
            grad_input = grad * \
                np.ones_like(self.parents[0].data) / self.parents[0].data.size
            grad_input = self.__match_shape(
                grad_input, self.parents[0].data.shape)
            self.parents[0].backward(grad_input)

        elif self.grad_fn == GradFn.EXP:
            grad_input = grad * self.data
            self.parents[0].backward(self.__match_shape(
                grad_input, self.parents[0].data.shape))

        elif self.grad_fn == GradFn.LOG:
            grad_input = grad / self.parents[0].data
            self.parents[0].backward(self.__match_shape(
                grad_input, self.parents[0].data.shape))

        elif self.grad_fn == GradFn.RELU:
            grad_input = grad * (self.parents[0].data > 0).astype(float)
            self.parents[0].backward(self.__match_shape(
                grad_input, self.parents[0].data.shape))

        elif self.grad_fn == GradFn.SIGMOID:
            sig = self.data
            grad_input = grad * sig * (1 - sig)
            self.parents[0].backward(self.__match_shape(
                grad_input, self.parents[0].data.shape))

    # ---------------- Debug ---------------- #
    def __repr__(self):
        return f"Tensor(data={self.data}, grad={self.grad})"


class SGD:
    def __init__(self, params, lr=0.01):
        self.params = params
        self.lr = lr

    def step(self):
        for p in self.params:
            if p.requires_grad and p.grad is not None:
                p.data -= self.lr * p.grad

    def zero_grad(self):
        for p in self.params:
            p.zero_grad()

File: wasm/tensor.go
package main

import (
	"fmt"
	"math"
)

// --- GradFn Enum --- //
type GradFn string

const (
	ADD  GradFn = "add"
	SUB  GradFn = "sub"
	MUL  GradFn = "mul"
	DIV  GradFn = "div"
	POW  GradFn = "pow"
	MEAN GradFn = "mean"
)

// --- Tensor --- //
type Tensor struct {
	Data         []float64
	RequiresGrad bool
	GradFn       *GradFn
	Grad         []float64
	Parents      []*Tensor
}

// makeTensor wraps data into a Tensor
func makeTensor(data []float64, requiresGrad bool) *Tensor {
	t := &Tensor{
		Data:         data,
		RequiresGrad: requiresGrad,
	}
	if requiresGrad {
		t.Grad = make([]float64, len(data))
	}
	return t
}

// Scalar helper
func scalar(v float64, requiresGrad bool) *Tensor {
	return makeTensor([]float64{v}, requiresGrad)
}

// matchShape (broadcast scalar to vector)
func (t *Tensor) matchShape(grad []float64) []float64 {
	// Case 1: Tensor is scalar → reduce grad to a single number
	if len(t.Data) == 1 && len(grad) > 1 {
		sum := 0.0
		for _, v := range grad {
			sum += v
		}
		return []float64{sum}
	}

	// Case 2: Broadcasting a scalar grad → expand
	if len(grad) == 1 && len(t.Data) > 1 {
		out := make([]float64, len(t.Data))
		for i := range out {
			out[i] = grad[0]
		}
		return out
	}

	// Case 3: Must match
	if len(grad) != len(t.Data) {
		panic(fmt.Sprintf("grad shape mismatch: got %d, expected %d", len(grad), len(t.Data)))
	}
	return grad
}

// --- Ops --- //

func (t *Tensor) Add(o *Tensor) *Tensor {
	data := make([]float64, max(len(t.Data), len(o.Data)))
	for i := range data {
		data[i] = t.Data[i%len(t.Data)] + o.Data[i%len(o.Data)]
	}
	fn := ADD
	return &Tensor{Data: data, RequiresGrad: true, GradFn: &fn, Parents: []*Tensor{t, o}}
}

func (t *Tensor) Sub(o *Tensor) *Tensor {
	data := make([]float64, max(len(t.Data), len(o.Data)))
	for i := range data {
		data[i] = t.Data[i%len(t.Data)] - o.Data[i%len(o.Data)]
	}
	fn := SUB
	return &Tensor{Data: data, RequiresGrad: true, GradFn: &fn, Parents: []*Tensor{t, o}}
}

func (t *Tensor) Mul(o *Tensor) *Tensor {
	data := make([]float64, max(len(t.Data), len(o.Data)))
	for i := range data {
		data[i] = t.Data[i%len(t.Data)] * o.Data[i%len(o.Data)]
	}
	fn := MUL
	return &Tensor{Data: data, RequiresGrad: true, GradFn: &fn, Parents: []*Tensor{t, o}}
}

func (t *Tensor) Pow(p *Tensor) *Tensor {
	data := make([]float64, len(t.Data))
	for i := range t.Data {
		data[i] = math.Pow(t.Data[i], p.Data[0]) // scalar exp for now
	}
	fn := POW
	return &Tensor{Data: data, RequiresGrad: true, GradFn: &fn, Parents: []*Tensor{t, p}}
}

func (t *Tensor) Mean() *Tensor {
	sum := 0.0
	for _, v := range t.Data {
		sum += v
	}
	mean := sum / float64(len(t.Data))
	fn := MEAN
	return &Tensor{Data: []float64{mean}, RequiresGrad: true, GradFn: &fn, Parents: []*Tensor{t}}
}

// --- Autograd --- //

func (t *Tensor) ZeroGrad() {
	if t.Grad != nil {
		for i := range t.Grad {
			t.Grad[i] = 0
		}
	}
}

func (t *Tensor) Backward(grad ...float64) {
	g := grad
	if len(g) == 0 {
		g = []float64{1.0}
	}
	g = t.matchShape(g)

	if t.RequiresGrad {
		if t.Grad == nil {
			t.Grad = make([]float64, len(t.Data))
		}
		for i := range t.Grad {
			t.Grad[i] += g[i]
		}
	}

	if t.GradFn == nil {
		return
	}

	switch *t.GradFn {
	case ADD:
		t.Parents[0].Backward(g...)
		t.Parents[1].Backward(g...)
	case SUB:
		t.Parents[0].Backward(g...)
		neg := make([]float64, len(g))
		for i, v := range g {
			neg[i] = -v
		}
		t.Parents[1].Backward(neg...)
	case MUL:
		a, b := t.Parents[0], t.Parents[1]
		gradA := make([]float64, len(a.Data))
		gradB := make([]float64, len(b.Data))
		for i := range g {
			gradA[i%len(a.Data)] += g[i] * b.Data[i%len(b.Data)]
			gradB[i%len(b.Data)] += g[i] * a.Data[i%len(a.Data)]
		}
		a.Backward(gradA...)
		b.Backward(gradB...)
	case POW:
		base, exp := t.Parents[0], t.Parents[1]
		gradBase := make([]float64, len(base.Data))
		for i := range base.Data {
			gradBase[i] = g[0] * exp.Data[0] * math.Pow(base.Data[i], exp.Data[0]-1)
		}
		base.Backward(gradBase...)
	case MEAN:
		x := t.Parents[0]
		gradX := make([]float64, len(x.Data))
		for i := range x.Data {
			gradX[i] = g[0] / float64(len(x.Data))
		}
		x.Backward(gradX...)
	}
}

// --- Optimizer --- //

type SGD struct {
	Params []*Tensor
	LR     float64
}

func (opt *SGD) Step() {
	for _, p := range opt.Params {
		if p.RequiresGrad && p.Grad != nil {
			for i := range p.Data {
				p.Data[i] -= opt.LR * p.Grad[i]
			}
		}
	}
}

func (opt *SGD) ZeroGrad() {
	for _, p := range opt.Params {
		p.ZeroGrad()
	}
}

// --- Helpers --- //

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// --- Main: Train linear regression --- //

// func main() {
// 	// training data: y = 2x + 5
// 	xs := make([]float64, 20)
// 	ys := make([]float64, 20)
// 	for i := 0; i < 20; i++ {
// 		xs[i] = -5 + float64(i)*(10.0/19.0)
// 		ys[i] = 2*xs[i] + 5
// 	}
// 	xTensor := makeTensor(xs, false)
// 	yTensor := makeTensor(ys, false)

// 	// init params
// 	w := scalar(rand.NormFloat64(), true)
// 	b := scalar(rand.NormFloat64(), true)
// 	optimizer := SGD{Params: []*Tensor{w, b}, LR: 0.01}

// 	for epoch := 0; epoch < 200; epoch++ {
// 		optimizer.ZeroGrad()

// 		preds := w.Mul(xTensor).Add(b)
// 		diff := preds.Sub(yTensor)
// 		loss := diff.Pow(scalar(2.0, false)).Mean()

// 		loss.Backward()
// 		optimizer.Step()

// 		if epoch%20 == 0 {
// 			fmt.Printf("Epoch %d: loss=%.4f, w=%.4f, b=%.4f\n",
// 				epoch, loss.Data[0], w.Data[0], b.Data[0])
// 		}
// 	}
// }
